### PR TITLE
Adding VPN options for IKE version and IKE split connections.

### PR DESCRIPTION
**Pull request by ggoodrich-ipp**
## Description

IKE version allows selecting ike (autoselect), ikev1, or ikev2.
Split connections gives an option of separating the first right subnet from the rest, and kicking out individual statements for each right subnet for better cross-compatibility.

Fixes: #3138 
## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?

We have been running this code in our production cloud for some months. Prior to this change, whenever an update was made to an existing customer gateway, it was wiping out the manual changes made to the configurations to split these out so that the gateway worked properly. Now, this code allows for these to be split into their own connections via the UI.

### CODE REVIEW

**Comment by rhtyd** [comment]
@ggoodrich-ipp please check and fix python/lint issues, see travis failure: https://travis-ci.org/github/apache/cloudstack/jobs/696639652

As PR has UI changes, please make submit a PR to the Primate repo [1] as per our agreement and vote to stop changes in the legacy UI [2] towards the next major release (4.15).

[1] https://github.com/apache/cloudstack-primate
[2] http://markmail.org/message/tblrbrtew6cvrusr

**Comment by rhtyd** [comment]
@ggoodrich-ipp np, that's why we've Travis. Yes, please send PR for Primate. 

@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@rhtyd @wido @weizhouapache do we merge this?

**Review by rhtyd — APPROVED** [review]
LGTM did not test it - I've concern if merging this near to the release may cause upgrade/use regressions

**Comment by DaanHoogland** [comment]
so do we move this to 4.16, @rhtyd ?

**Comment by weizhouapache** [comment]
@ggoodrich-ipp 
thanks for pr.
could you fix the conflicts and make primate ui change ?

**Comment by ggoodrich-ipp** [comment]
> @ggoodrich-ipp
> thanks for pr.
> could you fix the conflicts and make primate ui change ?

Should this be re-targeted to either 4.15 or 4.16 @DaanHoogland @rhtyd, I know the conflicts won't be resolved in master, due to the recent UI deprecation/switch to primate. We will be working towards getting these elements into a PR for primate soon, but it may be easier to include that in a separate PR. Thoughts?

**Comment by weizhouapache** [comment]
> > @ggoodrich-ipp
> > thanks for pr.
> > could you fix the conflicts and make primate ui change ?
> 
> Should this be re-targeted to either 4.15 or 4.16 @DaanHoogland @rhtyd, I know the conflicts won't be resolved in master, due to the recent UI deprecation/switch to primate. We will be working towards getting these elements into a PR for primate soon, but it may be easier to include that in a separate PR. Thoughts?

@ggoodrich-ipp 4.15.0.0 has been released, 4.15 branch will only accept bug fixes and security fixes afaik. this should be targeted to 4.16 I think.

**Comment by DaanHoogland** [comment]
@ggoodrich-ipp all the ui conflicts are to files that should now moved to ui/legacy. However I connot see them on github. (cc @rhtyd ?)

**Comment by ggoodrich-ipp** [comment]
> @ggoodrich-ipp all the ui conflicts are to files that should now moved to ui/legacy. However I connot see them on github. (cc @rhtyd ?)

I believe they've been removed permanently in the 4.16/master branch now. Maybe I need to redo this PR, and rip out the old UI parts, and instead rewrite for the new UI?

File: setup/db/create-schema.sql [hunk 1]
```diff
@@ -2280,6 +2280,8 @@ CREATE TABLE `cloud`.`s2s_customer_gateway` (
   `domain_id` bigint unsigned NOT NULL,
   `account_id` bigint unsigned NOT NULL,
   `removed` datetime COMMENT 'date removed if not null',
+  `ike_version` varchar(5) NOT NULL DEFAULT 'ike' COMMENT 'one of ike, ikev1, ikev2',
+  `split_connections` int(1) NOT NULL DEFAULT 0,
```
Comment (weizhouapache) [inline setup/db/create-schema.sql:2284]: @ggoodrich-ipp 
this affects only fresh installation.

need to add sqls in engine/schema/src/main/resources/META-INF/db/schema-41510to41600.sql
```
alter table s2s_customer_gateway add column `ike_version` varchar(5) NOT NULL DEFAULT 'ike' COMMENT 'one of ike, ikev1, ikev2';
alter table s2s_customer_gateway add column `split_connections` int(1) NOT NULL DEFAULT 0;
```

Comment (ggoodrich-ipp) [reply setup/db/create-schema.sql:2284]: Thanks @weizhouapache, we hadn't caught that. Those changes are now included.

**Comment by ggoodrich-ipp** [comment]
@DaanHoogland @rhtyd We've added the new UI elements, and removed the old UI elements. I believe we've resolved all conflicts and requests.

**Comment by weizhouapache** [comment]
@ggoodrich-ipp cool, thanks for update.
I think you need to revert the change below in setup/db/create-schema.sql

```
  `ike_version` varchar(5) NOT NULL DEFAULT 'ike' COMMENT 'one of ike, ikev1, ikev2',
  `split_connections` int(1) NOT NULL DEFAULT 0,
```

all travis test failed.

**Comment by ggoodrich-ipp** [comment]
> @ggoodrich-ipp cool, thanks for update.
> I think you need to revert the change below in setup/db/create-schema.sql
> 
> ```
>   `ike_version` varchar(5) NOT NULL DEFAULT 'ike' COMMENT 'one of ike, ikev1, ikev2',
>   `split_connections` int(1) NOT NULL DEFAULT 0,
> ```
> 
> all travis test failed with error
> 
> ```
> [ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.2.1:java (create-schema) on project cloud-developer: An exception occured while executing the Java class. null: InvocationTargetException: Unable to upgrade the database: Unable to execute upgrade script: Duplicate column name 'ike_version' -> [Help 1]
> ```

Shouldn't these SQL changes be in both places? I guess I'm confused. If you were doing a fresh installation, and not an upgrade, how would you get the columns?

**Comment by weizhouapache** [comment]
> > @ggoodrich-ipp cool, thanks for update.
> > I think you need to revert the change below in setup/db/create-schema.sql
> > ```
> >   `ike_version` varchar(5) NOT NULL DEFAULT 'ike' COMMENT 'one of ike, ikev1, ikev2',
> >   `split_connections` int(1) NOT NULL DEFAULT 0,
> > ```
> > 
> > 
> > all travis test failed with error
> > ```
> > [ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.2.1:java (create-schema) on project cloud-developer: An exception occured while executing the Java class. null: InvocationTargetException: Unable to upgrade the database: Unable to execute upgrade script: Duplicate column name 'ike_version' -> [Help 1]
> > ```
> 
> Shouldn't these SQL changes be in both places? I guess I'm confused. If you were doing a fresh installation, and not an upgrade, how would you get the columns?

@ggoodrich-ipp 
when setup a fresh installation, we need to setup database via cloudstack-setup-databases.
```
Using specified cluster management server node IP 10.135.122.40                 [ ^[[92mOK^[[0m ]
Checking Cloud database files ...                                               [ ^[[92mOK^[[0m ]
Checking local machine hostname ...                                             [ ^[[92mOK^[[0m ]
Checking SELinux setup ...                                                      [ ^[[92mOK^[[0m ]
Preparing /etc/cloudstack/management/db.properties                              [ ^[[92mOK^[[0m ]
Applying /usr/share/cloudstack-management/setup/create-database.sql             [ ^[[92mOK^[[0m ]
Applying /usr/share/cloudstack-management/setup/create-schema.sql               [ ^[[92mOK^[[0m ]
Applying /usr/share/cloudstack-management/setup/create-database-premium.sql     [ ^[[92mOK^[[0m ]
Applying /usr/share/cloudstack-management/setup/create-schema-premium.sql       [ ^[[92mOK^[[0m ]
Applying /usr/share/cloudstack-management/setup/server-setup.sql                [ ^[[92mOK^[[0m ]
Applying /usr/share/cloudstack-management/setup/templates.sql                   [ ^[[92mOK^[[0m ]
Processing encryption ...                                                       [ ^[[92mOK^[[0m ]
Finalizing setup ...                                                            [ ^[[92mOK^[[0m ]
```
it will create some tables.
when restart mgt server, cloudstack will perform database upgrade.
```
2021-02-16 14:58:47,736 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade40to41 to upgrade from 4.0.0-4.1.0 to 4.1.0
2021-02-16 14:58:55,307 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade410to420 to upgrade from 4.1.0-4.2.0 to 4.2.0
2021-02-16 14:59:05,234 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade420to421 to upgrade from 4.2.0-4.2.1 to 4.2.1
2021-02-16 14:59:05,348 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade421to430 to upgrade from 4.2.1-4.3.0 to 4.3.0
2021-02-16 14:59:10,576 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade430to440 to upgrade from 4.3.0-4.4.0 to 4.4.0
2021-02-16 14:59:17,909 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade440to441 to upgrade from 4.4.0-4.4.1 to 4.4.1
2021-02-16 14:59:18,179 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade441to442 to upgrade from 4.4.1-4.4.2 to 4.4.2
2021-02-16 14:59:18,203 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade442to450 to upgrade from 4.4.2-4.5.0 to 4.5.0
2021-02-16 14:59:24,759 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade443to444 to upgrade from 4.4.3-4.4.4 to 4.4.4
2021-02-16 14:59:24,792 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade450to451 to upgrade from 4.5.0-4.5.1 to 4.5.1
2021-02-16 14:59:24,917 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade451to452 to upgrade from 4.5.1-4.5.2 to 4.5.2
2021-02-16 14:59:25,149 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade452to453 to upgrade from 4.5.2-4.5.3 to 4.5.3
2021-02-16 14:59:25,173 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade453to460 to upgrade from 4.5.3-4.6.0 to 4.6.0
2021-02-16 14:59:26,139 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade460to461 to upgrade from 4.6.0-4.6.1 to 4.6.1
2021-02-16 14:59:26,180 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade461to470 to upgrade from 4.6.1-4.7.0 to 4.7.0
2021-02-16 14:59:26,608 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade470to471 to upgrade from 4.7.0-4.7.1 to 4.7.1
2021-02-16 14:59:26,716 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade471to480 to upgrade from 4.7.1-4.8.0 to 4.8.0
2021-02-16 14:59:26,848 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade480to481 to upgrade from 4.8.0-4.8.1 to 4.8.1
2021-02-16 14:59:26,884 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade481to490 to upgrade from 4.8.1-4.9.0 to 4.9.0
2021-02-16 14:59:29,859 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade490to4910 to upgrade from 4.9.0-4.9.1.0 to 4.9.1.0
2021-02-16 14:59:30,392 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade4910to4920 to upgrade from 4.9.1.0-4.9.2.0 to 4.9.2.0
2021-02-16 14:59:30,681 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade4920to4930 to upgrade from 4.9.2.0-4.9.3.0 to 4.9.3.0
2021-02-16 14:59:30,797 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade4930to41000 to upgrade from 4.9.3.0-4.10.0.0 to 4.10.0.0
2021-02-16 14:59:32,061 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade41000to41100 to upgrade from 4.10.0.0-4.11.0.0 to 4.11.0.0
2021-02-16 14:59:33,903 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade41100to41110 to upgrade from 4.11.0.0-4.11.1.0 to 4.11.1.0
2021-02-16 14:59:34,096 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade41110to41120 to upgrade from 4.11.1.0-4.11.2.0 to 4.11.2.0
2021-02-16 14:59:34,320 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade41120to41130 to upgrade from 4.11.2.0-4.11.3.0 to 4.11.3.0
2021-02-16 14:59:34,348 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade41120to41200 to upgrade from 4.11.2.0-4.12.0.0 to 4.12.0.0
2021-02-16 14:59:35,463 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade41200to41300 to upgrade from 4.12.0.0-4.13.0.0 to 4.13.0.0
2021-02-16 14:59:36,613 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade41300to41310 to upgrade from 4.13.0.0-4.13.1.0 to 4.13.1.0
2021-02-16 14:59:36,635 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade41310to41400 to upgrade from 4.13.1.0-4.14.0.0 to 4.14.0.0
2021-02-16 14:59:37,499 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade41400to41500 to upgrade from 4.14.0.0-4.15.0.0 to 4.15.0.0
2021-02-16 14:59:39,493 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade41500to41510 to upgrade from 4.15.0.0-4.15.1.0 to 4.15.1.0
2021-02-16 14:59:39,529 DEBUG [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Running upgrade Upgrade41510to41600 to upgrade from 4.15.1.0-4.16.0.0 to 4.16.0.0
```


hope it is clear to you.

**Comment by ggoodrich-ipp** [comment]
> hope it is clear to you.
@weizhouapache Yes, thanks for the explanation. I didn't realize that a new install was upgrading through all the versions. That makes sense now. Pushed.

**Comment by dhansenippathways** [comment]
It looks like the build is failing on a timeout error.  Should we get the build re-kicked off or are there other errors that I'm missing?

@DaanHoogland @ggoodrich-ipp

**Comment by DaanHoogland** [comment]
don't worry @dhansenippathways , these sometimes happen because we are to heavy for travis. I restarted those and we test in other ways as well.
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by dhansenippathways** [comment]
Let me know if there's anything else I need to do on this pull request (i'm helping out @ggoodrich-ipp with this).

**Comment by weizhouapache** [comment]
code lgtm

**Review by weizhouapache — APPROVED** [review]

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
not sure about the static rules failure but the life cycle failures are due to centos7 limitation which are addressed in #4727 amongst many other things, This can be easily ported to a separate PR if we need to; https://github.com/apache/cloudstack/pull/4727/commits/8eec8c6cf6a6784ee7bc24a50114414d33b8b698

**Comment by weizhouapache** [comment]
> not sure about the static rules failure but the life cycle failures are due to centos7 limitation which are addressed in #4727 amongst many other things, This can be easily ported to a separate PR if we need to; [8eec8c6](https://github.com/apache/cloudstack/commit/8eec8c6cf6a6784ee7bc24a50114414d33b8b698)

@DaanHoogland 
but it worked before. might it be caused by a recent commit ? need to solve it asap.

**Comment by DaanHoogland** [comment]
@weizhouapache Centos7 does not support live migration like ubuntu does due to the default installation of qemu. I think this default must have changesd recently but it was never stable. My fix would mean checking the host os and skipping the test if it is `Centos`. I would merge this with the live migration tests failing on centos. I'll start an ubuntu test.
@blueorangutan test centos7 kvm-ubuntu

**Comment by rhtyd** [comment]
@ggoodrich-ipp can you fix the conflicts?

**Comment by rhtyd** [comment]
@weizhouapache @GabrielBrascher @wido @kiwiflyer @ravening @soreana @svenvogel @DaanHoogland @andrijapanicsb @shwstppr @Pearl1594 @Spaceman1984 - this is an enhancement that may benefit other users - should we be considering for 4.15.1 iff it can be reviewed and tested?

**Comment by soreana** [comment]
@rhtyd @ggoodrich-ipp  How I can help to test this pr? Could you provide some test scenario?

**Comment by DaanHoogland** [comment]
@rhtyd @soreana @ggoodrich-ipp, I backported to 4.15 branch in #4904

please have a look and a spin. cc @borisstoyanov @weizhouapache

**Comment by DaanHoogland** [comment]
@dhansenippathways can you help test #4595

**Comment by andrijapanicsb** [comment]
This needs some serious testing by someone who has been playing and/or suffering with VPNs. Let's make sure we don't break anything. I like the PR in general.

4.16 would be a better candidate, time-wise

**Comment by ggoodrich-ipp** [comment]
> @dhansenippathways can you help test #4595

We tested this, and have been running these changes in our Production environment for around one year (4.11.3). We recently added the new UI elements to satisfy the requirements of the current versions.

I'm hesitant to resolve the conflicts, as they're in a file that will continue to receive conflicts if the PR is not merged in a timely manner. I will certainly resolve them once it is determine which branch this PR will get pulled into. Thanks!

**Comment by rhtyd** [comment]
Thanks @ggoodrich-ipp good to know this is already tested in production, @DaanHoogland thanks let's review/test on https://github.com/apache/cloudstack/pull/4904
@ggoodrich-ipp should we close this PR then, and can you help share your testing/feedback details on https://github.com/apache/cloudstack/pull/4904

**Comment by ggoodrich-ipp** [comment]
> Thanks @ggoodrich-ipp good to know this is already tested in production, @DaanHoogland thanks let's review/test on #4904
> @ggoodrich-ipp should we close this PR then, and can you help share your testing/feedback details on #4904

I don't have strong feelings on which PR we choose, as long as we get this enhancement into the code base so that at some point in the future it is there. I will let you @rhtyd @DaanHoogland @weizhouapache and others make that determination. We are already managing this change in our own fork for our production environment. Thanks!

**Comment by rhtyd** [comment]
Moved to https://github.com/apache/cloudstack/pull/4953 and fixed

**Pull request closed without merging**